File: client/src/main/js/components/CourseLabels.js
import React from 'react';
import {Badge} from 'reactstrap';
import {Seq} from 'immutable';

export function CourseLabels(props) {
    const statusLabel =
        <Badge className="course-label" color="primary">{props.course.state.lifecycle.toLowerCase()}</Badge>;

    const techLabels =
        Seq([
            props.course.language,
            props.course.testingLanguage,
            props.course.testingFramework
        ])
            .map(value => value.toLowerCase())
            .map(value => <Badge className="course-label" color="info">{value}</Badge>);

    const servicesLabels =
        props.course.state
            .activatedServices
            .map(value => value.toLowerCase())
            .map(service => <Badge className="course-label" color="warning">{service}</Badge>);

    return (
        <span className="course-labels">
            {statusLabel}
            {servicesLabels}
            {techLabels}
        </span>
    );
}

File: frontend/src/main/kotlin/org/flaxo/frontend/component/courseLabels.kt
package org.flaxo.frontend.component

import org.flaxo.common.data.Course
import react.RBuilder
import react.dom.div
import react.dom.span

/**
 * Adds course labels.
 */
fun RBuilder.courseLabels(course: Course) = div {
    span(classes = "course-labels") {
        statusLabel(course)
        servicesLabels(course)
        techLabels(course)
    }
}

private fun RBuilder.servicesLabels(course: Course) =
        course.state.activatedServices
                .map { it.name }
                .map { it.toLowerCase() }
                .forEach { span(classes = "course-label badge badge-warning") { +it } }

private fun RBuilder.techLabels(course: Course) =
        setOf(course.settings.language, course.settings.testingLanguage, course.settings.testingFramework)
                .filterNotNull()
                .map { it.toLowerCase() }
                .forEach { span(classes = "course-label badge btn-info") { +it } }

private fun RBuilder.statusLabel(course: Course) =
        span(classes = "course-label badge btn-primary") {
            +course.state.lifecycle.name.toLowerCase()
        }
